# IV.12.8. e)  МАКАРОВА МАРИЯ 

In [7]:
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt
import pylab
from sympy import diff, symbols, cos, sin, exp , ln, sqrt
from sympy import *
x, y = symbols('x y')


In [77]:
def f(x):
    m = sqrt(x)*exp(-x)
    return(m)   


In [21]:
y=sqrt(x)*exp(-x)

dif_text=y.diff(x)

dif_text
#производная

-sqrt(x)*exp(-x) + exp(-x)/(2*sqrt(x))

In [16]:
          
k=solve(dif)
#нуль производной 

[1/2]

In [70]:
def dif(x):
    d= exp(-x)*(-sqrt(x)+1/(sqrt(4*x)))
    return(d) 
half = dif(1/2) 
half
#полувысота

-6.73384303554051e-17

Функция  sqrt(x)^exp(-x) = −6.73384303554051⋅10−17

    Для левого и правого корня возьму разные функции для итераций.
    g1 = (exp(-1+2x))/ 8      <---- производная меньше 1 при х < 1.193, поэтому эта функция для поиска левого корня
    g2 = ln(sqrt(8 * e * x))  <---- производная меньше 1 при х > 0.5 , поэтому эта функция для поиска правого корня

In [71]:
def g1(x):
    t =(exp(-1+2*x))/ 8
    return(t)

def g2(x):
    t= ln(sqrt(8 * exp(1) * x))
    return(t)

def q (x0,x1,x2):
    q1 = abs((x2 - x1)/(x1-x0))
    return(q1)
def eps(x1,x2,q):
    e = (abs(x2-x1))/(1-q)
    return(e)

In [73]:
w1=1/2
w0 = p = 1000000
i = 0
epss = 1000000000
while(epss>=0.001):
    w2 = g1(w1)
    
    if(i>0):
        qq = q(w0,w1,w2)
        epss = eps(w1,w2,qq)
        print ("Эпсилон :")  
        print(round(epss,4))
        w1=p
        w1=w2
        w0=p
       
        print( i , "-тая итерация")
        print(" ")  
    i = i+1    
print(" ")    
print ("Итоговое эпсилон :")    
print(round(epss,4))
print(" ") 
www1 = round(w2,3)
print(" ")    
print ("x1 =", www1)  

    

Эпсилон :
0.3750
1 -тая итерация
 
Эпсилон :
0.0660
2 -тая итерация
 
Эпсилон :
0.0073
3 -тая итерация
 
Эпсилон :
0.0007
4 -тая итерация
 
 
Итоговое эпсилон :
0.0007
 
 
x1 = 0.051


In [74]:
w1=1/2
w0 = p = 1000000
i = 0
epss = 1000000000
while(epss>=0.001):
    w2 = g2(w1)
    
    if(i>0):
        qq = q(w0,w1,w2)
        epss = eps(w1,w2,qq)
        print ("Эпсилон :")  
        print(round(epss,4))
        w1=p
        w1=w2
        w0=p
       
        print( i , "-тая итерация")
        print(" ")  
    i = i+1    
print(" ")    
print ("Итоговое эпсилон :")    
print(round(epss,4))
print(" ") 
www2 = round(w2,3)
print(" ")    
print ("x2 =", www2)  

Эпсилон :
0.6931
1 -тая итерация
 
Эпсилон :
0.4349
2 -тая итерация
 
Эпсилон :
0.1554
3 -тая итерация
 
Эпсилон :
0.0456
4 -тая итерация
 
Эпсилон :
0.0126
5 -тая итерация
 
Эпсилон :
0.0034
6 -тая итерация
 
Эпсилон :
0.0009
7 -тая итерация
 
 
Итоговое эпсилон :
0.0009
 
 
x2 = 1.846


In [75]:
 print ( "Определенная методом простой итерации ширина функции на полувысоте с точностью 10^3:" , www2-www1)

Определенная методом простой итерации ширина функции на полувысоте с точностью 10^3: 1.795
